### DISCUSSION TITLE
calling readMessage twice doesn't work in AWS lambda

### DISCUSSION THREAD

**Discussion by sasa130** [openpgpjs/openpgpjs#1635] [Q&A]
Hi, I'm running to an issue where readMessage just seems to fail silently when called twice after decrypting 2 different private keys.  

1st Scenario:
1. Decrypt latest private key (different from the one used to encrypt the file) - Success
2. Call readMessage function - returns error: Session key decryption failed.   
3. Call decrypt function - Success
Lambda continues execution with logic to retry with another key
4. Decrypt previous version of private key (same key used to encrypt the file) - Success
5. Call readMessage function - No error was logged but lambda simply finishes execution.

I thought it had something to do with the way I've written the code but I copied the exact code in a unit test and it properly logged the result/error and finishes the execution as expected.  To be sure, I tried the scenarios below:

2nd Scenario:
1. Decrypt previous version of private key (same key used to encrypt the file) - Success
2. Call readMessage function - continues on successfully
3. Call decrypt function - Success
(Modified lambda to continue execution even if the decryption was already successful)
4. Decrypt latest private key (different from the one used to encrypt the file) - Success
5. Call readMessage function - No error was logged but lambda simply finishes execution, not continuing with the call to decrypt function.  

I tried the same thing with my unit test and it also logged the result/error and terminates at the exp  finishes the execution as expected.  

3rd Scenario:
Variation of the previous ones except on both times, I used the correct private key. The result is the same, the second call to readMessage function just silently finishes and ends the execution of the lambda even though it should have successfully continued and decrypted the file.

It seems then there's a problem calling readMessage more than once.  Any thoughts?

**Comment by twiss** [comment]
Hmm, weird. Maybe there's some issue with the event loop in AWS lambda, such that it thinks the function / promise has completed, while it hasn't? The promise chain inside OpenPGP.js can be fairly complex. I don't know why it would break on the second call, though, unless it has something to do with the ponyfill loading - are you using streams?

As a workaround, you could call `readMessage` only once, and then check which key should be used to decrypt it using `message.getEncryptionKeyIDs()`, and then decrypt only that private key, and use that to decrypt the file. That might be more efficient, anyway, as then you don't need to parse the message multiple times.

**Reply by sasa130** [reply]
Thanks @twiss for the suggestion, that sounds good and I'll definitely try that out.  I also made the code as simple as possible by simply awaiting the call to the async readMessage function within the same function so the issue can't be further confused with anything else but the issue persists.

**Reply by larabr** [reply]
@sasa130 Are you maybe passing the same stream as input to readMessage?
If not, then could you share your code?

**Reply by sasa130** [reply]
Hi @larabr , yes, I am passing the same original stream (s3 object) on both occasions to the readObject.  Could you please explain why would that have been the problem if it is?  

And @twiss , your suggestion worked, thanks!  Called readMessage once and then compared the encryptionKeyIds first of the message against that of the private key before calling the decrypt function.

**Reply by larabr** [reply]
@sasa130 the `readMessage` function consumes the input stream (as it's processed by the decrypt function), so passing the same stream again won't work as it doesn't contain the [read] data anymore

**Reply by kahlan-dev** [reply]
Gotcha, thanks @larabr, @twiss  .  Calling readMessage once and checking the right encryptionkey first works.
